tensorstore/kvstore/s3/aws_metadata_credential_provider.h

```diff
@@ -0,0 +1,47 @@
+// Copyright 2023 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TENSORSTORE_KVSTORE_S3_AWS_METADATA_CREDENTIAL_PROVIDER_H
+#define TENSORSTORE_KVSTORE_S3_AWS_METADATA_CREDENTIAL_PROVIDER_H
+
+#include <memory>
+
+#include "tensorstore/internal/http/http_transport.h"
+#include "tensorstore/kvstore/s3/aws_credential_provider.h"
+#include "tensorstore/util/str_cat.h"
+#include "tensorstore/util/result.h"
+
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+/// Provides S3 credentials from the EC2 Metadata server
+/// if running within AWS
+class EC2MetadataCredentialProvider : public AwsCredentialProvider {
+ private:
+  std::shared_ptr<internal_http::HttpTransport> transport_;
+
+ public:
+  EC2MetadataCredentialProvider(
+      std::shared_ptr<internal_http::HttpTransport> transport)
+      : transport_(std::move(transport)) {}
+
+  Result<AwsCredentials> GetCredentials() override;
```

Comment:
It seems like we may want to convert this from `Result<AwsCredentials> GetCredentials()` to `Future<AwsCredentials> GetCredentials()`.

Reply:
Those are separate issues, but perhaps.  Also, defer that work if you want until we get this done.

Reply:
> It seems like we may want to convert this from `Result<AwsCredentials> GetCredentials()` to `Future<AwsCredentials> GetCredentials()`.

I guess this so that the `GetCredentials` here can be converted into chained futures so as to not block the thread on which futures are submitted?